### PR TITLE
Adding the new muon track finder index and three eta cuts features in the GT emulator

### PR DESCRIPTION
**Pull request by elfontan**
#### PR description:
This PR includes two updates of the GT emulator accorging to the latest version of the utm library v0.11.2 (see details in [CMSLITDPG-1104](https://its.cern.ch/jira/browse/CMSLITDPG-1104)) and developments of the L1 menu for Run 3:
* Usage of three eta cuts (up to five cuts are allowed) in the design of the muon algorithms with the upt requirement. The check on the eta range is now implemented as a function checking a vector of eta windows to handle the five cuts in a easier way.
* New track finder index feature for muons in the uGT emulator needed for the update of the 2023 L1 menu as detailed in [CMSLITDPG-1075](https://its.cern.ch/jira/browse/CMSLITDPG-1075)). Note that the missing muon index data is also included in the GtRecordDumper, needed to produce a test vector as input to the firmware validation. The bits assigned to muon index in the muon object are 36..42 (see Table 6 in the uGT firmware documentation - [scales_inputs_2_ugt.pdf](https://github.com/cms-l1-globaltrigger/mp7_ugt_legacy/blob/master/doc/scales_inputs_2_ugt/pdf/scales_inputs_2_ugt.pdf)). Thank you @arnobaer and Herbert for the work!

These updates are needed for the deployment of the Run L1 menu [L1Menu_Collisions2023_v1_0_1](https://github.com/cms-l1-dpg/L1MenuRun3/tree/master/development/L1Menu_Collisions2023_v1_0_1) in view of the first stable beams at 13.6 TeV or soon after. 
Backport to CMSSW_13_0_X: https://github.com/cms-sw/cmssw/pull/41313.

#### PR validation:
Basic tests performed successfully starting from CMSSW_13_1_X_2023-04-06-1100
> cmsrel CMSSW_13_1_X_2023-04-06-1100
> cd CMSSW_13_1_X_2023-04-06-1100/src
> cmsenv
> git cms-addpkg L1Trigger/L1TGlobal 
> git cms-addpkg L1Trigger/Configuration
> scram b distclean 
> git cms-checkdeps -a -A
> scram b -j 8
> scram b runtests 
> scram build code-checks
> scram build code-format

A cross-check of the countings of each trigger bit in the new firmware and in the updated emulator showed a perfect match.

Basic cmsDriver command for L1Ntuple production successfully tested:
> cmsDriver.py l1Ntuple -s RAW2DIGI --python_filename=test.py -n 1000 --no_output --era=Run3 --data --conditions=124X_dataRun3_Prompt_v4  --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW --customise=L1Trigger/L1TNtuples/customiseL1Ntuple.L1NtupleRAWEMU --filein=/store/data/Run2022F/EphemeralZeroBias0/RAW/v1/000/361/468/00000/52351179-2329-47d8-bffc-a01833bb1704.root --fileout=L1Ntuple.root

Note that the temporary release CMSSW_13_1_X_2023-04-06-1100 is not available anymore, so that final tests have been performed in CMSSW_13_1_X_2023-04-15-1100.

### CODE REVIEW

File: L1Trigger/L1TGlobal/interface/ConditionEvaluation.h [hunk 1]
```diff
@@ -342,7 +350,35 @@ namespace l1t {
                            << "\n\t passWindow2 = " << passWindow2
                            << "\n\t pass W1 || W2 = " << (passWindow1 || passWindow2) << std::endl;
 
-    if (passWindow1 || passWindow2) {
+    if (W3beginR == W2endR && W3beginR == (Type1)-1) {
```
Comment (aloeliger) [inline L1Trigger/L1TGlobal/interface/ConditionEvaluation.h:353]: Does this templated check require that `Type1` be of some type with a definable subtraction with integers? Is that guaranteed?

**Comment by elfontan** [comment]
Hi @aloeliger,
honestly I followed here the logic that was used for checking the validity until now (both for eta & phi and explained [here](https://github.com/elfontan/cmssw/blob/804ac2031000a64182604c1650602758ddb5c9bf/L1Trigger/L1TGlobal/interface/ConditionEvaluation.h#L277-L288 )) without any further investigation. I am not sure why the chosen implementation was this one, but until now it was always found to be in agreement with the firmware behaviour and did not show any problem.
(Maybe @arnobaer or @rekovic or @cavana, who were around before me, know some details?) 
Otherwise do you have in mind any additional test to further validate this @aloeliger?

Cheers,
--Elisa

P.S. I am waiting for the confirmation about the firmware and emulator agreement before moving the PR to ready for review.

**Comment by aloeliger** [comment]
please test

**Comment by elfontan** [comment]
Dear all,
after a first round of validation of the GT firmware and emulator behaviour with @arnobaer, the PR has been updated to fix some mismatches in the L1_SingleJet*HE countings caused by the updated checkRangeEta code. While investigating this issue, the code has been improved to implement the check on the eta range as a function that checks a vector of eta windows. In this way, the possibility to have up to five eta cuts is handled in an easier way.

In parallel, an issue with the firmware related to the BMTF muon monitoring seeds based on the usage of the new muon TF index feature has been spotted. We wait for the fix before moving to another round of validation and for now we keep this PR on hold.

We'll keep you posted. Thank you,
--Elisa

**Comment by elfontan** [comment]
Hello @aloeliger @epalencia all,
I am moving the PR to Ready for review given that the firmare bug was fixed and the firmware-emulator agreement was validated (see [CMSLITDPG-1102](https://its.cern.ch/jira/browse/CMSLITDPG-1102)).
Is any action needed on the commit history to proceed?

Cheers,
--Elisa

**Comment by aloeliger** [comment]
@elfontan Is a new version of the firmware necessary in the externals for emulator agreement or function?

**Comment by elfontan** [comment]
No no, it was an internal bug in the uGT firmware logic itself (no utm-related). 
The countings for the index feature were different from the emulator results because of this bug (always zero) so we couldn't double-check the 100% correct implementation of the current emulator before the bug fix.
(@arnobaer can correct me if I am wrong)

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
+l1

File: L1Trigger/L1TGlobal/src/MuonTemplate.cc [hunk 2]
```diff
@@ -101,10 +101,17 @@ void MuonTemplate::print(std::ostream& myCout) const {
...
-    myCout << "    etaWindow1Lower   =" << std::hex << m_objectParameter[i].etaWindow1Lower << std::endl;
-    myCout << "    etaWindow1Upper   =" << std::hex << m_objectParameter[i].etaWindow1Upper << std::endl;
-    myCout << "    etaWindow2Lower   =" << std::hex << m_objectParameter[i].etaWindow2Lower << std::endl;
-    myCout << "    etaWindow2Upper   =" << std::hex << m_objectParameter[i].etaWindow2Upper << std::endl;
+
+    size_t etaWindowIndex = 1;
+    for (const auto& window : m_objectParameter[i].etaWindows) {
+      //myCout << "    etaWindows[" << etaWindowIndex << "].lower =" << std::hex << window.lower << std::endl;
+      //myCout << "    etaWindows[" << etaWindowIndex << "].upper =" << std::hex << window.upper << std::endl;
+      std::cout << " # MUON TEMPLATE:    etaWindows[" << etaWindowIndex << "].lower =" << std::hex << window.lower
```
Comment (perrotta) [inline L1Trigger/L1TGlobal/src/MuonTemplate.cc:109]: Why not `myCout` here? Was it done on purpose?

Comment (elfontan) [reply L1Trigger/L1TGlobal/src/MuonTemplate.cc:109]: Thanks Andrea, that's a bad leftover from some debugging...

**Comment by perrotta** [comment]
+1
- Last commit only fixed some debug statement
- I assume that the previous signature of @aloeliger for @cms-sw/l1-l2 can be still considered as good

**Pull request merged**